dojo_plugin/api/v1/dojo.py

```diff
@@ -0,0 +1,244 @@
+            return {"success": False, "error": f"Invalid dojo specified: {data.get('dojo_id')}"}
+
+        if not (is_admin() or any(award.name == "PUBLIC_DOJO" for award in user.awards)):
+            return {
+                "success": False,
+                "error":
+                """<p><b>You do not have authorization to make public dojos.</b></p>"""
+                """<p>For authorization to make public dojos, please email """
+                """<a href="mailto:pwn-college@asu.edu">pwn-college@asu.edu</a>.</p>"""
+            }
+
+        dojo.join_code = None
+        db.session.add(dojo)
+        db.session.commit()
+        return {"success": True, "dojo_id": dojo.id}
+
+
+@dojo_namespace.route("/delete")
+class DeleteDojo(Resource):
+    @authed_only
+    def post(self):
+        data = request.get_json()
+        user = get_current_user()
+
+        dojo_id = data.get("dojo_id")
+        dojo = Dojos.query.filter_by(id=dojo_id).first()
+        if not is_dojo_admin(user, dojo):
+            return {"success": False, "error": f"Invalid dojo specified: {data.get('dojo_id')}"}
+
+        dojo_dir = DOJOS_DIR/str(user.id)/dojo.id
+        if not dojo_dir.exists():
+            return {"success": False, "error": "Dojo directory does not exist."}
+
+        db.session.delete(dojo)
+        db.session.commit()
+
+        shutil.rmtree(str(dojo_dir))
+
+        return {"success": True, "dojo_id": dojo.id}
+
+@dojo_namespace.route("/create")
+class CreateDojo(Resource):
+    @authed_only
+    def post(self):
+        data = request.get_json()
+        user = get_current_user()
+
+        with tempfile.TemporaryDirectory(dir=DOJOS_DIR, prefix=str(user.id), suffix=".git-clone") as tmp_dir:
+            try:
+                dojo_repo = data.get("dojo_repo")
+                GIT_SSH_REGEX = "^git@github.com:[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$"
+                GIT_HTTPS_REGEX = "^https://github.com/[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$"
+                assert re.match(GIT_SSH_REGEX, dojo_repo) or re.match(GIT_HTTPS_REGEX, dojo_repo), (
+                    f"Repository violates regular expression. Must match <code>{GIT_SSH_REGEX}</code> or <code>{GIT_HTTPS_REGEX}</code>."
+                )
+
+                # clone it!
+                clone_dir = pathlib.Path(tmp_dir)/"clone"
+                clone_dir.mkdir()
+                returncode, output = sandboxed_git_clone(dojo_repo, str(ctfd_to_host_path(clone_dir)))
+
+                N=b"\n"
+                assert returncode == 0, (
+                    f"Dojo clone failed with error code {returncode}:<br>"
+                    f"<code>{output.replace(N,b'<br>').decode('latin1')}</code><br>"
+                    "Please make sure that you properly added the deploy key to the repository settings, "
+                    "and properly entered the repository URL."
+                )
+
+                # figure out the dojo ID
+                dojo_specs = list(clone_dir.glob("*.yml"))
```

Comment:
Maybe we shouldn't offer such flexibility. Let's just check for `dojo.yml`. In the future we might want other ymls.

Reply:
There is no other way to specify dojo IDs right now, so we either DOJO_ID.yml or get the ID somehow from the yml (whether it's ID-ifying the dojo name or having an ID field there).

In general, there are probably massive issues if the dojo ID changes in an update, btw. Challenges will be unable to be found and will get recreated, with no solves. I don't check for that, but it's something to keep in mind.